sklearn/impute/_knn.py

```diff
@@ -0,0 +1,264 @@
+        Distance metric for searching neighbors. Possible values:
+        - 'nan_euclidean'
+        - callable : a user-defined function which conforms to the
+          definition of _pairwise_callable(X, Y, metric, **kwds). The function
+          accepts two arrays, X and Y, and a ``missing_values`` keyword in
+          ``kwds`` and returns a scalar distance value.
+
+    copy : boolean, optional(default=True)
+        If True, a copy of X will be created. If False, imputation will
+        be done in-place whenever possible.
+
+    Attributes
+    ----------
+    statistics_ : array, shape (n_features, )
+        The 1-D array contains the mean of each feature calculated using
+        observed (i.e. non-missing) values. This is used for imputing
+        missing values in samples that are either excluded from nearest
+        neighbors search because they have too many missing features or
+        all of the sample's k-nearest neighbors (i.e., the potential donors)
+        also have the relevant feature value missing.
+
+    References
+    ----------
+    * Olga Troyanskaya, Michael Cantor, Gavin Sherlock, Pat Brown, Trevor
+      Hastie, Robert Tibshirani, David Botstein and Russ B. Altman, Missing
+      value estimation methods for DNA microarrays, BIOINFORMATICS Vol. 17
+      no. 6, 2001 Pages 520-525.
+
+    Examples
+    --------
+    >>> from sklearn.impute import KNNImputer
+    >>> nan = float("NaN")
+    >>> X = [[1, 2, nan], [3, 4, 3], [nan, 6, 5], [8, 8, 7]]
+    >>> imputer = KNNImputer(n_neighbors=2)
+    >>> imputer.fit_transform(X)
+    array([[1. , 2. , 4. ],
+           [3. , 4. , 3. ],
+           [5.5, 6. , 5. ],
+           [8. , 8. , 7. ]])
+    """
+
+    def __init__(self, missing_values=np.nan, n_neighbors=5,
+                 weights="uniform", metric="nan_euclidean", copy=True):
+
+        self.missing_values = missing_values
+        self.n_neighbors = n_neighbors
+        self.weights = weights
+        self.metric = metric
+        self.copy = copy
+
+    def _calc_impute(self, dist, receivers_idx, X_col, fit_X_col,
+                     non_missing_fix_X_col, statistic):
+        """Helper function to calculate the imputation value of a single
+        column.
+
+        Parameters
+        ----------
+        dist : array-like, shape = (n_receivers, n_train_samples, )
+            distance matrix between the receivers and the train samples
+
+        receivers_idx : array-like, shape = (n_receivers, )
+            indices of X_col to be imputed
+
+        X_col : array-like, shape = (n_samples, )
+            column to be imputed
+
+        fit_X_col : array-like, shape = (n_train_samples, )
+            column from training
+
+        non_missing_fix_X_col : array-like, shape = (n_train_samples, )
+            mask for missing values corresponding to ``fit_X_col``
+
+        statistic : float
+            statistic for column generated from fitting ``fit_X_col``
```

Comment:
Doesn't document ``Returns``.

Reply:
Does the shape/return type depend on the input? That's kinda funky

Reply:
Refactored this to be less funky.